README.md
# view-3d-model

[![NPM Package][npm]][npm-url]
[![NPM Downloads][npm-downloads-per-month]][npm-trends]

**_view-3d-model_** is a `Vue.js` library that makes use of `three.js` and `GLTFLoader` to allow users to display 3D models in their `Vue.js` applications.

This can be achieved by importing the `ThreeDModel` component from the library, which supports models in `gltf` format, as well as the binary version of `gltf`, which is `glb`.

```
npm install view-3d-model
```

```html
<script setup>
  import ThreeDModel from "view-3d-model";
</script>

<template>
  <div class="wrapper">
    <ThreeDModel
      class="three-d-model"
      file-path="./models/your-model-name.glb"
      :use-editor="true"
      :custom-settings="{
      autoRotate: true
    }"
    />
  </div>
</template>

<style>
  .wrapper {
    height: 100vh;
    width: 100vw;
  }

  .three-d-model {
    height: 100%;
    width: 100%;
  }
</style>
```

## Chapters

This documentation includes the following chapters:

<details>
<summary>Get Started</summary>

- [Get Started](#get-started)
  - [Prerequisites](#pre-requisites)
  - [Start using **_view-3d-model_**](#start-using-view-3d-model)
  - [Create a folder for your models](#create-folder-for-models)
  - [Using `glb`](#using-glb)
  - [Using `gltf`](#using-gltf)
  - [Install **_view-3d-model_**](#install-view-3d-model)
  - [Import and use the `ThreeDModel` in Vue.js projects](#import-and-use-vue)
  - [Import and use the `ThreeDModel` in Nuxt.js projects](#import-and-use-nuxt)

</details>

<details>
<summary>Using ThreeDModel</summary>

- [Using `ThreeDModel`](#using-three-d-model)
  - [Short on props in Vue.js](#short-on-props)

</details>

<details>
<summary>Customize `ThreeDModel` with props</summary>

- [Customize `ThreeDModel` with props](#customize-three-d-model-with-props)
  - [filePath](#file-path)
  - [customSettings](#custom-settings)
  - [useEditor](#use-editor)

</details>

<details>

<summary>Control `ThreeDModel` with CSS</summary>

- [Control `ThreeDModel` with CSS](#control-three-d-model-with-css)
  - [Set width and height](#set-width-and-height)
  - [Set background](#set-background)

</details>

<details>
<summary>Extensions</summary>

- [Extensions](#extensions)
  - [Supported extensions](#supported-extensions)

</details>

<details>
<summary>Troubleshooting</summary>

- [TroubleShooting](#trouble-shooting)
  - [My 3d model doesn't show up](#model-not-showing-up)
  - [I can't set the lights](#cant-set-lights)
  - [My model is not looking correct](#not-looking-correct)

</details>

<details>
<summary>Create a Vue.js Project</summary>

- [Create a Vue.js Project](#create-vue-project)
  - [Follow these steps to create a Vue.js application:](#steps-to-create-vue)

</details>

<details>
<summary>ThreeDModel</summary>

- [ThreeDModel](#three-d-model)
  - [Props](#props)
  - [Events](#events)
  - [Methods](#methods)

</details>

## <a id="get-started"></a> Get Started

### <a id="pre-requisites"></a> Prerequisites

To use **_view-3d-model_** you will need a `Vue.js` project (or a `Nuxt` project), and a `gltf(glb)` file.

- If you are unfamiliar to the `Vue.js` javascript framework,
  [see this guide to create a Vue.js project](#create-vue-project)

- If you don't have a `gltf(glb)` file, there are plenty of free downloads on the web. For example you can visit [Sketchfab](https://sketchfab.com/features/free-3d-models), and download a 3d-model of your liking.
  Make sure that you choose the file format `gltf` or `glb`.

:information_source: <b>The `glb` version usually is about 30% smaller than the `gltf` version and thus recommended to use if possible.</b>

### <a id="start-using-view-3d-model"></a> Start using **_view-3d-model_**

When you have a `glb` or `gltf` file, and a `Vue.js` project (or a `Nuxt` project), follow these steps to start using **_view-3d-model_**:

### <a id="create-folder-for-models"></a> Create a folder for your models

- Lets create a folder called `models` inside the `public` folder.

:information_source: <b>If you are using `Nuxt` use the `static` folder instead of `public`.</b>

### <a id="using-glb"></a> Using `glb`

- If you have a model of the format `glb`, you can simply add the file to the `models` folder as it is:

```
your-project/
...
    public/                              // Use static folder if using Nuxt
        models/
            name-of-your-model.glb
...
```

### <a id="using-gltf"></a> Using `gltf`

- If you have a model of the format `gltf`, you'll notice that it consists of a number of different items. Usually there is a `texture folder` or a number of image files, a `bin` file, a `license.txt` file and a `gltf` file. Create a folder with the name of your model inside the `models` folder, and add all the contents of the `gltf` file to this folder.

Your project should look something like this:

```
your-project/
...
    public/                             // Use static folder if using Nuxt
        models/
            name-of-your-model/
                textures
                license.txt
                scene.bin
                scene.gltf

...

```

### <a id="install-view-3d-model"></a> Install **_view-3d-model_**

- Install **_view-3d-model_** by typing the following in the command line:

```
npm install view-3d-model
```

### <a id="import-and-use-vue"></a> Import and use the `ThreeDModel` component in Vue.js projects

**_[Click here to see how to import and use `ThreeDModel` in Nuxt project](#import-and-use-nuxt)_**

Follow these steps to import and use the `ThreeDModel` component in your `Vue.js` project:

In `App.vue`, or any other component where you want to use `ThreeDModel`:

Using `Composition API (Vue 3)`:

```html
<script setup>
  // import the ThreeDModel component from view-3d-model
  import ThreeDModel from "view-3d-model";
</script>

<template>
  <div class="wrapper">
    <!-- If using glb format -->
    <ThreeDModel
      class="three-d-model"
      file-path="./models/your-model-name.glb"
    />

    <!-- If using gltf format -->
    <ThreeDModel
      class="three-d-model"
      file-path="./models/your-model-name/scene.gltf"
    />
  </div>
</template>
<!-- In this example we set the width and height to match the viewport. -->
<style>
  .wrapper {
    height: 100vh;
    width: 100vw;
  }

  .three-d-model {
    height: 100%;
    width: 100%;
  }
</style>
```

Using `Options API (Vue 2)`:

```html
<template>
  <div class="wrapper">
    <!-- If using glb format -->
    <ThreeDModel
      class="three-d-model"
      file-path="./models/your-model-name.glb"
    />

    <!-- If using gltf format -->
    <ThreeDModel
      class="three-d-model"
      file-path="./models/your-model-name/scene.gltf"
    />
  </div>
</template>

<script>
  // import the ThreeDModel component from view-3d-model
  import ThreeDModel from "view-3d-model";
  export default {
    name: "nameOfYourComponent",
    // Declare ThreeDModel in component options
    components: {
      ThreeDModel,
    },
  };
</script>
<!-- In this example we set the width and height to match the viewport. -->
<style>
  .wrapper {
    height: 100vh;
    width: 100vw;
  }

  .three-d-model {
    height: 100%;
    width: 100%;
  }
</style>
```

To make `ThreeDModel` available globally throughout your `Vue.js` project add this to `main.js`:

```js title="src/main.js"
import { createApp } from "vue";
import App from "./App.vue";
// Import ThreeDModel from view-3d-model
import ThreeDModel from "view-3d-model";

createApp(App)
  // Register ThreeDModel as global component
  .component("ThreeDModel", ThreeDModel)
  .mount("#app");
```

Now you can use `ThreeDModel` wherever you want in your `Vue.js` project without importing it in your components.
**_[Click here to see how to start using `ThreeDModel`](#using-three-d-model)_**

### <a id="import-and-use-nuxt"></a> Import and use the `ThreeDModel` component in Nuxt.js projects

- If you haven't already done so, install **_view-3d-model_** by typing the following in the command line:

```
npm install view-3d-model
```

- In the `plugins` folder create a file called `view-3d-model.js`.

- Copy and paste the following into `plugins/view-3d-model.js`:

```js title="view-3d-model.js"
import Vue from "vue";
import ThreeDModel from "view-3d-model";

Vue.component("ThreeDModel", {
  extends: ThreeDModel,
});
```

- In `nuxt.config.js`, add the following to the `plugins` array:

```js title="nuxt.config.js"
...
    plugins: [
        ...
            {src: '~/plugins/view-3d-model.js', mode: 'client'}
        ...
    ]
...
```

- Also in `nuxt.config.js`, add **_view-3d-model_** to `build.transpile`:

```js title="nuxt.config.js"
...
  build: {
    ...
      transpile: ['view-3d-model']
    ...
  }
...
```

Now `ThreeDModel` is available globally throughout your `Nuxt` project without importing it in your components like so:

```html
<template>
    <!-- If using glb format -->
    <client-only>
      <div class="wrapper">
        <ThreeDModel
          class="three-d-model"
          file-path="./models/your-model-name.glb"
        />
      </div>
    </client-only>

    <!-- If using gltf format -->
    <client-only>
      <div class="wrapper">
        <ThreeDModel
          class="three-d-model"
          file-path="./models/your-model-name/scene.gltf"
        />
    </client-only>

</template>

<script>
  export default {
    name: "nameOfYourComponent",
  };
</script>
<!-- In this example we set the width and height to match the viewport. -->
<style>
  .wrapper {
    height: 100vh;
    width: 100vw;
  }

  .three-d-model {
    height: 100%;
    width: 100%;
  }
</style>
```

## <a id="using-three-d-model"></a> Using `ThreeDModel`

By now you should have a 3d model rendered to the screen. Now let's take a look at how you can adjust the settings of the model via props.

### <a id="short-on-props"></a> Short on props in Vue.js

- `ThreeDModel` takes three props: `filePath`, `useEditor` and `customSettings`. When we're passing props to a Vue component we typically do so in `kebab-case`, so in this guide we will be following that convention.
  (As in the examples above when we're passing the `filePath` prop, we do so like this: `file-path="./models/your-model-name.glb"`).

- When we want to send a dynamic prop value, a javascript expression or a number rather than a string, we will use the `:` `v-bind` shortcut like so:

```html
<ThreeDModel
  class="three-d-model"
  file-path="./models/your-model-name.glb"
  :use-editor="true"
/>
```

Here we're passing the prop `useEditor` using v-bind which means the value will be the `Boolean` value `true`.
Without the `:` the value would be a `String` value 'true'.

## <a id="customize-three-d-model-with-props"></a> Customize `ThreeDModel` with props

`ThreeDModel` takes three props:

- ### <a id="file-path"></a> `filePath (String)`
  We've already used the required prop `filePath` which is the path that points at your 3d-model to be loaded.
- ### <a id="custom-settings"></a> `customSettings (Object)`:
  An object that specifies settings to camera, lighting, orbit controls and rotation.
  The default values looks like this:

```js
{
//  Field of view, defines the extent of the scene that is seen on the display. Set in degrees, defaults to 50.
//  type: Number, min: 1, max: 180
   fov: 50,

 // Camera position, will be used if x, y and z is provided. Else it will be ignored and the camera position will be set automatically.
//  type: Number,  min: -50, max: 50
   cameraPosition: {
     x: null,
     y: null,
     z: null,
   },

  // Exposure level, defaults to 1
  // type: Number, min: 0, max: 10
  exposure: 1,

 // A light in a specific direction, used to simulate daylight
 // Position is set to x:0, y:1, z:0 meaning that the light shines from the top down.
 // Set intensity and color adjust the light
   directionalLight: {
    // type: Number, min: 0, max: 100
     intensity: 0.5,
    // Recommended: Hex("FFFFFF") or Dec(16777215)
     color: "#FFFFFF",
   },

 // Illuminates the scene equally, without direction, set intensity and color to adjust the light
   ambientLight: {
    // type: Number, min: 0, max: 100
     intensity: 0.1,
    //Recommended: Hex("#FFFFFF") or Dec(16777215)
     color: "#FFFFFF",
   },

 // If set to true enables to interact with the model (zoom, grab, rotate), defaults to true
 // type: Boolean
 enableOrbitControls: true,

 // If set to true rotates the model, defaults to false
 // Type: Boolean
 autoRotate: false,

 // The rotation speed if autoRotate is set to true, defaults to 2
 // type: Number, min: 0.1, max: 100
 rotationSpeed: 2,

}
```

You can pass the `customSettings` prop to adjust all or some of the values like in this example:

:information_source: <b>Note that when passing values to the `cameraPosition` all of the properties `x`, `y` and `z` are required. If one or two of these values are left out, `ThreeDModel` will fall back to its default behaviour and set the camera position automatically</b>

```html
<ThreeDModel
  class="three-d-model"
  file-path="./models/your-model-name.glb"
  :use-editor="true"
  :custom-settings="{
    fov: 70,
    directionalLight: {
      intensity: 1
    },
    autoRotate: true
  }"
/>
```

Here we increase the `fov` value to 70, raises the intensity of the `directionalLight` to 1, and sets the `autoRotate` to true.

To make it easier to set the values of the `customSettings` prop, it's recommended to use the editor by setting prop `useEditor` to `true`.

:information_sourse: <b>Note that if the model you are using was created using the extension `KHR_materials_unlit`, no lights will be applicable, and thus no controls for `directionalLight` and `ambientLight` will show up in the editor. [read more about extensions here](#extensions) </b>

- ### <a id="use-editor"></a> `useEditor` (`Boolean`)
  When we set the value of prop `useEditor` to `true`, an editor will be created. This editor lets you play around and adjust the settings of the camera, the lighting, orbit controls and rotation.

If you want to use the current settings there are two options:

:information_source: <b>Note that the prop 'useEditor' will not be included when using `Copy as Template` or `Use Settings`</b>

- Click the `Copy as Template` button, which will copy a `ThreeDModel` template to the clipboard that you can paste into your project.
- Click the `Use Settings` button. This will `$emit` an event with the current settings wich makes it possible to listen to the event in your parent component like so:

Using `composition API (Vue 3)`:

```html
<script setup>
  ...
    function handleSettings(settings) {
      console.log('current settings: ', settings)
    }
  ...
</script>

<template>
  ...
  <ThreeDModel
    class="three-d-model"
    file-path="./models/your-model-name.glb"
    :use-editor="true"
    @use-settings="data => handleSettings(data)"
  />
  ...
</template>
```

Using `Options API (Vue 2)`:

```html
<template>
  ...
  <ThreeDModel
    class="three-d-model"
    file-path="./models/your-model-name.glb"
    :use-editor="true"
    @use-settings="data => handleSettings(data)"
  />
  ...
</template>

<script>
  export default {
    ...
      methods: {
        handleSettings(settings) {
          console.log('current settings: ', settings)
        }
      }
  ...
  };
</script>
```

The function `handleSettings` will be called everytime you click the `useSettings` button in the editor. If you want save the settings in a database or similar, just put your logic for this in the handleSettings function.

## <a id="control-three-d-model-with-css"></a> Control `ThreeDModel` with CSS

Here you can read about how to use CSS to control the 3d model

### <a id="set-width-and-height">Set width and height</a>

- In order to work correctly the `ThreeDmodel` component needs a width and height. If not, the width and height properties of the rendered canvas element that holds your model will be 0 and thus your model won't be visible. It's recommended to wrap the `ThreeDModel` in a div, and set the width and height of `ThreeDModel` relative to the wrapper div:

```html
<template>
  <div class="wrapper">
    <ThreeDModel
      class="three-d-model"
      file-path="./models/your-model-name.glb"
    />
  </div>
</template>

<!-- In this example we set the width and height to match the viewport. -->
<style>
  .wrapper {
    height: 100vh;
    width: 100vw;
  }

  .three-d-model {
    height: 100%;
    width: 100%;
  }
</style>
```

We can use the CSS property `aspect-ratio` to define the ratio between width and height of the wrapper:

```html
<template>
  <div class="wrapper">
    <ThreeDModel
      class="three-d-model"
      file-path="./models/your-model-name.glb"
    />
  </div>
</template>
<!-- In this example we set the width of the wrapper to 80vw and the aspect-ratio to 4/3. -->
<style>
  .wrapper {
    width: 80vw;
    aspect-ratio: 4/3;
  }

  .three-d-model {
    height: 100%;
    width: 100%;
  }
</style>
```

If we want we can use media querys to set the size of the wrapper:

```html
<template>
  <div class="wrapper">
    <ThreeDModel
      class="three-d-model"
      file-path="./models/your-model-name.glb"
    />
  </div>
</template>
<!-- In this example we use a media query breakpoint to change the width on larger devices -->
<style>
  .wrapper {
    width: 100vw;
    aspect-ratio: 4 / 3;
  }

  @media screen and (min-width: 992px) {
    .wrapper {
      width: 50vw;
    }
  }

  .three-d-model {
    height: 100%;
    width: 100%;
  }
</style>
```

By default `ThreeDModel` will center the model and set the the camera position. [Here you can read about how to adjust the camera position and other settings](#customize-three-d-model-with-props).

### <a id="set-background"></a> Set background

To set the background color you simply do so by setting it either to the `ThreeDModel` itself or it's parent like so:

```html
<template>
  <div class="wrapper">
    <ThreeDModel
      class="three-d-model"
      file-path="./models/your-model-name.glb"
    />
  </div>
</template>
<style>
  .wrapper {
    background-color: #29bdc1;
    width: 80vw;
    aspect-ratio: 4/3;
  }

  .three-d-model {
    height: 100%;
    width: 100%;
  }
</style>
```

Or you can use a background image like in this example:

```html
<template>
  <div class="wrapper">
    <ThreeDModel
      class="three-d-model"
      file-path="./models/your-model-name.glb"
    />
  </div>
</template>

<style>
  .wrapper {
    width: 100vw;
    height: 100vh;
    background-image: url("https://images.pexels.com/photos/1242348/pexels-photo-1242348.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=1");
    background-repeat: no-repeat;
    background-position: center;
    background-size: cover;
  }

  .three-d-model {
    width: 100%;
    height: 100%;
  }
</style>
```

## <a id="extensions"></a> Extensions

When creating a gltf file there are a number of extensions one can use.
**_view-3d-model_** is using `three.js` to load and render 3d models, and thus supports the same extensions as the `GLTFLoader` in the `three.js` library.

While the most common extensions out there are supported, sometimes trying to load a file with a an unknown extensions can lead to problems.

One example of this worth mentioning is the deprecated extension `KHR_materials_pbrSpecularGlossiness)`.
Support for this extensions in `three.js` was dropped in november 2022, and while most gltfs already use the workflow `metal/rough` instead there still are some downloads using the extension. To solve this you'll need to convert the file to `metal/rough`. If your model is using this extension you will get a console message with a link to [this article where you can read about how to convert `spec/gloss` to `metal/rough`](https://www.donmccurdy.com/2022/11/28/converting-gltf-pbr-materials-from-specgloss-to-metalrough/).

Another common case which is not a problem but worth mentioning to avoid confusion, is when the gltf was created using the extension `KHR_materials_unlit`. When so, no lights will be applicable to the model and hence no controls for `ambientLight` or `directionalLight` will be created in the editor. [read more about the `KHR_materials_unlit extension` here](https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_materials_unlit/README.md).

### <a id="supported-extensions"></a> Supported extensions:

- KHR_draco_mesh_compression
- KHR_materials_clearcoat
- KHR_materials_ior
- KHR_materials_specular
- KHR_materials_transmission
- KHR_materials_iridescence
- KHR_materials_unlit
- KHR_materials_volume
- KHR_mesh_quantization
- KHR_lights_punctual1
- KHR_texture_basisu
- KHR_texture_transform
- EXT_texture_webp
- EXT_meshopt_compression
- EXT_mesh_gpu_instancing


CHANGELOG.md
## [1.3.1] 2023-05-12

### Added

- Update docs

## [1.3.0] 2023-05-11

### Added

- Added damping to `orbitControls` to give a sense of weight when interacting with the model.

## [1.2.4] 2023-05-10

### Fixed

- Replaced computed property `currentSettings` with method `getCurrentsettings` to prevent reactivity bug.

## [1.2.3] 2023-05-06

### Removed

- Unused data property `containerDimensions`.

## [1.2.2] 2023-05-06

### Changed

- Split up code for readabilility

## [1.2.1] 2023-04-24

### Removed

- Removed unnecessary watcher for container dimensions. Call `updateCameraAndRenderer` directly from `onWindowResize` instead.

### Changed

- Moved styles to `style.css`

## [1.2.0] 2023-04-23

### Added

- `exposure` to `customSettings`

- New chapter to docs: Control `ThreeDModel` with CSS

## [1.1.4] 2023-04-21

### Fixed

- Prevent duplicated class names in copied template.

### Changed

- 'Copied to clipboard' msg instead of alert when clicking Copy as Template button.

## [1.1.3] 2023-04-20

### Added

- Update docs

## [1.1.2]

### Added 2023-04-20

- Update docs

## [1.1.1] 2023-04-19

### Fixed

- Disabled eslint in `custom-logs.js` and `validate-props.js`

## [1.1.0] 2023-04-19

### Added

- Checking if extension `KHR_materials_unlit` is being used in the gltf.
  If so, no light intensity or color settings will be applicable.
  No inputs for these values are created in the editor.

- Checking if extension `KHR_materials_pbrSpecularGlossiness` is being used.
  Since the extension is not supported a warning message will show in the console.

### Fixed

- Prevented customized console logs when running in 'production' mode.

### Changed

- Moved validation to `utils/validate-props`

## [1.0.6] - 2023-04-18

### Added

- Documentation
- `filePath` to object in $emit event `useSettings`

### Fixed

- Issue with loader animation

### Changed

- Name of editor button `Copy as Component` to `Copy as Template`

## [1.0.5] - 2023-04-17

### Added

- Documentation, more to come

## [1.0.4] - 2023-04-15

### Fixed

- Update package.json

## [1.0.3] - 2023-04-15

### Fixed

- Issues with missing files

## [1.0.2] - 2023-04-15

### Added

- Validation for prop `customSettings`.
- Customized log messages `msg`, `warn` & `error`.
- `reset` button in editor.

### Changed

- Name of `$emit` event `clicked`. Now called `useSettings`.
- Folder structure in editor.

## [1.0.1] - 2023-04-13

### Added

- Loader animation when loading model.

### Fixed

- Not rendering component if loading failes.

### Changed

- Only importing necessary components from three.js library

## [1.0.0] - 2023-04-11

### Added

- Component `ThreeDModel`, for rendering gltf(glb) 3d models.

<!--
Template

### Added

### Fixed

### Changed

### Deprecated

### Removed

-->


src/mixins/methods.js
/* eslint-disable */

import {
  Scene as Scene,
  PerspectiveCamera as PerspectiveCamera,
  WebGLRenderer as WebGLRenderer,
  sRGBEncoding as sRGBEncoding,
  LinearToneMapping as LinearToneMapping,
  Box3 as Box3,
  Vector3 as Vector3,
  AmbientLight as AmbientLight,
  DirectionalLight as DirectionalLight,
  PMREMGenerator as PMREMGenerator,
  REVISION as REVISION,
} from "three";

import { GLTFLoader as GLTFLoader } from "three/examples/jsm/loaders/GLTFLoader.js";
import { DRACOLoader as DRACOLoader } from "three/examples/jsm/loaders/DRACOLoader";
import { OrbitControls as OrbitControls } from "three/examples/jsm/controls/OrbitControls";
import { RoomEnvironment as RoomEnvironment } from "three/examples/jsm/environments/RoomEnvironment.js";
import { GUI as GUI } from "lil-gui";

import log from "../utils/view-3d-model-logs.js";

export default {
  name: "methods",
  methods: {
    // Creating scene, camera and renderer
    initialize() {
      this.scene = new Scene();
      this.camera = new PerspectiveCamera(
        this.fov,
        this.$el.clientWidth / this.$el.clientHeight,
        this.near,
        this.far
      );
      const container = this.$el;
      this.renderer = new WebGLRenderer({
        container,
        antialias: true,
        alpha: true,
      });
      this.renderer.setPixelRatio(window.devicePixelRatio);
      this.renderer.setSize(this.$el.clientWidth, this.$el.clientHeight);
      this.renderer.outputEncoding = sRGBEncoding;
      this.renderer.toneMapping = LinearToneMapping;
      this.renderer.toneMappingExposure = this.exposure;
      this.$el.appendChild(this.renderer.domElement);

      window.addEventListener("resize", () => this.onWindowResize(), false);
    },

    // Adding lights to the scene
    // Using values from props 'ambientLighting' and 'directionalLigthing'
    setLights() {
      // Setting ambient light color and intensity
      this.ambientLight = new AmbientLight(
        this.ambientLighting?.color,
        this.ambientLighting?.intensity
      );
      this.scene.add(this.ambientLight);

      // Setting directional light color and intensity
      this.directionalLight = new DirectionalLight(
        this.directionalLighting?.color,
        this.directionalLighting?.intensity
      );

      this.scene.add(this.directionalLight);
    },

    // Adding animation loop
    animate() {
      window.requestAnimationFrame(this.animate);
      this.render();
      if (this.orbitControls) {
        this.orbitControls.update();
      }
    },

    // Adding an environment to the scene, to provide an even lighting to the scene
    setEnvironment() {
      this.pmremGenerator = new PMREMGenerator(this.renderer);
      this.pmremGenerator.compileEquirectangularShader();
      this.environment = this.pmremGenerator.fromScene(
        new RoomEnvironment()
      ).texture;

      this.scene.environment = this.environment;
    },

    // Creating gltfLoader
    // Creating dracoLoader which is used if gltf is compressed with draco
    createLoader() {
      this.gltfLoader = new GLTFLoader();
      this.gltfLoader.setCrossOrigin("anonymous");
      this.dracoLoader = new DRACOLoader();
      this.dracoLoader.setDecoderPath(
        `https://unpkg.com/three@0.${REVISION}.x/examples/jsm/libs/draco/gltf/`
      );
      this.gltfLoader.setDRACOLoader(this.dracoLoader);
    },

    // Setting isUnlit to true if method/usingUnllit returns true
    // If so, no lights will be applicable
    checkExtensions(extensions, extensionsUsed) {
      this.isUnlit = this.usingUnlit(extensions, extensionsUsed);
      this.isUnlit &&
        log.msg(
          "This model was created using extension 'KHR_materials_unlit', hence no lights will be applicable. Read more about 'KHR_materials_unlit: ' https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_materials_unlit/README.md"
        );

      // Logs a warning if method/usingSpecGloss returns true
      if (this.usingSpecGloss(extensions, extensionsUsed)) {
        log.warn(
          "This model was created using extension 'KHR_materials_pbrSpecularGlossiness'. This extension is not supported and hence the model might not appear correct. Read more about 'KHR_materials_pbrSpecularGlossiness': https://www.donmccurdy.com/2022/11/28/converting-gltf-pbr-materials-from-specgloss-to-metalrough/"
        );
      }
    },

    // Returns true or false depending on if extension 'KHR_materials_unlit' is used.
    usingUnlit(extensions, extensionsUsed) {
      return !!(
        extensions?.KHR_materials_unlit ||
        extensionsUsed?.includes("KHR_materials_unlit")
      );
    },

    // Returns true or false depending on if extension 'KHR_materials_pbrSpecularGlossiness' is used.
    usingSpecGloss(extensions, extensionsUsed) {
      return !!(
        extensions?.KHR_materials_pbrSpecularGlossiness ||
        extensionsUsed?.includes("KHR_materials_pbrSpecularGlossiness")
      );
    },

    // Creating a bounding box from model and returns the size and center
    getCenterAndSize() {
      const box = new Box3().setFromObject(this.model);
      const size = box.getSize(new Vector3()).length();
      const center = box.getCenter(new Vector3());
      return { size, center };
    },

    // Centering the loaded model
    centerModel(center) {
      this.model.position.x += this.model.position.x - center.x;
      this.model.position.y += this.model.position.y - center.y;
      this.model.position.z += this.model.position.z - center.z;
    },

    // Setting the camera position
    setCamera(size, center) {
      this.camera.near = size / 100;
      this.camera.far = size * 100;
      this.updateCamera();

      // Using camera position values if provided in customSettings
      if (
        this.cameraPosition.x &&
        this.cameraPosition.y &&
        this.cameraPosition.z
      ) {
        this.camera.position.x = this.cameraPosition.x;
        this.camera.position.y = this.cameraPosition.y;
        this.camera.position.z = this.cameraPosition.z;
      }

      // Else using size to set camera position (This is the default behavior)
      else {
        this.camera.position.x += size / 1.5;
        this.camera.position.y += size / 3;
        this.camera.position.z += size / 1.5;
      }
      // Point camera at center of the scene
      this.camera.lookAt(center);
    },

    // Adding orbit controls to be able to interact with the model, enabled if enableOrbitControls is set to true
    setControls(size) {
      this.orbitControls = new OrbitControls(
        this.camera,
        this.renderer.domElement
      );
      this.orbitControls.enabled = this.enableOrbitControls;
      this.orbitControls.enableDamping = true;
      this.orbitControls.dampingFactor = 0.06;
      this.orbitControls.maxDistance = size * 10;
    },

    // Adding rotation if autoRotate is set to true and rotation speed
    setRotation() {
      this.orbitControls.autoRotate = this.autoRotate;
      this.orbitControls.autoRotateSpeed = this.rotationSpeed;
    },

    // Updating camera projectionMatrix
    updateCamera() {
      this.camera.updateProjectionMatrix();
    },

    // Renders the scene and camera
    render() {
      this.renderer.render(this.scene, this.camera);
    },

    // Updating camera aspect ratio, projectionMatrix and renderers size
    updateCameraAndRenderer() {
      this.camera.aspect = this.$el.clientWidth / this.$el.clientHeight;
      this.updateCamera();
      this.renderer.setSize(this.$el.clientWidth, this.$el.clientHeight);
    },

    // Runs updateCameraAndRenderer on window resize
    onWindowResize() {
      this.updateCameraAndRenderer();
    },

    // Returns the current settings att any given moment.
    // Will be used when emitting or copying settings from editor
    getCurrentSettings() {
      return {
        fov: this.camera.fov,
        cameraPosition: {
          x: this.camera.position.x,
          y: this.camera.position.y,
          z: this.camera.position.z,
        },
        exposure: this.renderer.toneMappingExposure,
        directionalLight: {
          intensity: this.directionalLight.intensity,
          color: this.directionalLight.color,
        },
        ambientLight: {
          intensity: this.ambientLight.intensity,
          color: this.ambientLight.color,
        },
        enableOrbitControls: this.orbitControls.enabled,
        autoRotate: this.orbitControls.autoRotate,
        rotationSpeed: this.orbitControls.autoRotateSpeed,
      };
    },

    // Emitting current settings
    handleUseSettings() {
      this.$emit("useSettings", {
        filePath: this.filePath,
        customSettings: this.getCurrentSettings(),
      });
    },

    // Copies a ThreeDModel template with current settings to clipboard
    async handleCopy() {
      const settings = JSON.stringify(this.getCurrentSettings()).replaceAll(
        '"',
        ""
      );
      const classList = this.$el.classList.value;
      const component = `
      <ThreeDModel
      class="${classList.replace("model-wrapper", "")}"
      file-path="${this.filePath}"
      :custom-settings="${settings}"/>`;
      try {
        await navigator.clipboard.writeText(component);
        this.triggerCopyMsg();
      } catch (err) {
        alert("failed to copy");
      }
    },

    // Setting 'showCopyMsg' to true for one second
    triggerCopyMsg() {
      this.showCopyMsg = true;
      setTimeout(() => {
        this.showCopyMsg = false;
      }, 1000);
    },

    // Resets all values in Editor
    resetEditor() {
      this.gui.reset();
    },

    // Creating GUI to help adjusting the model
    createEditor() {
      this.gui = new GUI();
      this.gui.title("Editor");

      // Camera folder
      const cameraFolder = this.gui.addFolder("Camera").close();
      // Fov
      cameraFolder
        .add(this.camera, "fov", 1, 180)
        .name("fov")
        .onChange(this.updateCamera);

      // Camera position folder
      const cameraPositionFolder = cameraFolder
        .addFolder("Camera Position")
        .close();

      // Camera position x
      cameraPositionFolder.add(this.camera.position, "x", -50, 50).listen();
      // Camera position y
      cameraPositionFolder.add(this.camera.position, "y", -50, 50).listen();
      // Camera position z
      cameraPositionFolder.add(this.camera.position, "z", -50, 50).listen();

      // Lights folder
      const lightsFolder = this.gui.addFolder("Light").close();

      // Exposure
      lightsFolder
        .add(this.renderer, "toneMappingExposure", 0, 10)
        .name("Exposure");

      // Creating controls for directionalLight and ambientLight if not using extension 'KHR_materials_unlit'
      if (!this.isUnlit) {
        // Directional light folder
        const directionalLightFolder = lightsFolder
          .addFolder("Directional Light")
          .close();
        // Intensity
        directionalLightFolder.add(this.directionalLight, "intensity", 0, 100);
        // Color
        directionalLightFolder.addColor(this.directionalLight, "color");

        // Ambient light folder
        const ambientLightFolder = lightsFolder
          .addFolder("Ambient Light")
          .close();
        // Intensity
        ambientLightFolder.add(this.ambientLight, "intensity", 0, 100);
        // Color
        ambientLightFolder.addColor(this.ambientLight, "color");
      }

      // Orbit controls folder
      const orbitControlsFolder = this.gui.addFolder("Orbit Controls").close();

      // Enable/Disable Orbit controls
      orbitControlsFolder
        .add(this.orbitControls, "enabled")
        .name("Enable Orbit Controls");

      // Rotation folder
      const rotationFolder = this.gui.addFolder("Rotation").close();

      // Enable/Disable Auto Rotate
      rotationFolder.add(this.orbitControls, "autoRotate").name("Auto Rotate");

      // Auto Rotation Speed
      rotationFolder
        .add(this.orbitControls, "autoRotateSpeed", 0.1, 100)
        .name("Rotation Speed");

      // Buttons
      // Reset button, runs
      this.gui.add(this, "resetEditor").name("Reset");
      // Copy button, runs method/handleCopy when clicked
      this.gui.add(this, "handleCopy").name("Copy as Template");
      // Emit button, runs method/handleEmit when clicked
      this.gui.add(this, "handleUseSettings").name("Use Settings");
    },
  },
};
